zio-http/src/main/scala/zio/http/Route.scala

```diff
@@ -0,0 +1,503 @@
+package zio.http
+
+import io.netty.handler.codec.http.HttpHeaderNames
+import zio._
+import zio.http.model.{Headers, MediaType, Status}
+import zio.http.socket.{SocketApp, WebSocketChannelEvent}
+import zio.stream.ZStream
+
+import java.io.{File, FileNotFoundException}
+import java.nio.file.Paths
+import java.util.zip.ZipFile
+
+sealed trait Route[-R, +Err, -In, +Out] { self =>
```

Comment:
This is a "polishing" question:

The name `Route` is not exactly accurate, it's more like `RoutingLogic`, which is too weird and long to be the right name here.

Given the name is an approximation anyway, could it make sense to instead call this `Http`? 

The rational would be:

1. Fewer changes required in the many libraries that already use ZIO HTTP.
2. It's a little nice for the "entry point" data type in a library to have a name that is similar to the library's name (e.g. `ZIO` in ZIO Core).

Thoughts?

Reply:
I was thinking the same but wanted to keep a separate new name while the new model is being discussed. (Although in general I don’t think that’s a very good name for what this trait does either :)

Reply:
The question is what do we call the related types then. For example the Http that is “really” http (request to response), and the http vs handler middleware’s

Reply:
Makes sense to me. An `Http` is a program, that, given an input, either handles it with a handler, or chooses not to handle it.

Reply:
How about this proposal:

 - `Http` &mdash; a program that, when given an input, either handles it with a handler, or chooses not to handle it. Mental model: `Set[(RoutePattern, RouteHandler)]` (my personal favorite), or `Route => Option[RouteHandler]` (classically).
 - `Handler` &mdash; Handles an input to produce an output.
 - `HttpAspect` &mdash; Transforms an `Http`.
 - `HandlerAspect` &mdash; Transforms a `Handler`.

This doesn't allow reuse of `Middleware` terminology at all, which is unfortunate. But it would be really weird to use `Middleware` for `Handler` and not `Http`, yet the vast majority of aspects are `Handler` transformers, not `Http` transformers.

By the way, another possibility occurs to me: `Route` (`Http`) already allows application of `HandlerAspect`, because there is a way to lift any `HandlerAspect` into an `HttpAspect`: by choosing some subset of handlers on which to apply the aspect (currently: "forall").

I think that means `HandlerAspect` could extend `HttpAspect`.

That means that `myApp @@ <xyz>` can work for any `<xyz>`, be they handler aspects or http aspects. That, in turn, implies we could use the `Middleware` name for `HttpAspect`, and mix into the companion object not merely all of the `HttpAspect` (which extend or contract the table size) but all of the `HandlerAspect`. 

In this alternative:

 - `Http` &mdash; a program that, when given an input, either handles it with a handler, or chooses not to handle it. Mental model: `Set[(RoutePattern, RouteHandler)]` (my personal favorite), or `Route => Option[RouteHandler]` (classically).
 - `Handler` &mdash; Handles an input to produce an output.
 - `Middleware` &mdash; Transforms an `Http`. Companion object includes constructor for ALL `Middleware`, including those in companion object of `HandlerAspect`.
 - `HandlerAspect` &mdash; Transforms a `Handler`. Extends `Middleware` using "for all" semantic.

Thoughts?

Reply:
I think I like the second. The idea of applying the aspect to all handlers indicates how potentially we could support the world where the `Http` is not a "flat" tree, basically the table has nesting in your table analogy. That seems like it could be useful. Like you want to apply some logic to all the `users` handlers which could be done very cleanly if all the users routes can themselves be their own table which is then embedded into a larger table.